File: frontend/lib/api/business.ts
import { getApiBaseUrl } from "@/lib/api/client";

async function headersJson(token: string) {
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
  } as const;
}

async function parseJson<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    if (text) {
      try {
        const json = JSON.parse(text) as Record<string, unknown>;
        if (typeof json.detail === "string") throw new Error(json.detail);
        if (typeof json.message === "string") throw new Error(json.message);
      } catch (e) {
        if (e instanceof Error && e.message !== text) throw e;
      }
    }
    throw new Error(text || `${res.status}`);
  }
  return res.json() as Promise<T>;
}

export type BusinessWorkspace = {
  workspace_id: string;
  title: string;
  business_type: string;
  total_budget: string | number | null;
  currency: string;
  created_by: string;
  status: string;
  created_at?: string | null;
  updated_at?: string | null;
};

export type BusinessUnit = {
  unit_id: string;
  workspace_id: string;
  name: string;
  unit_type: string;
  location: string | null;
  manager_user_id: string | null;
  budget_limit: string | number | null;
  status: string;
  created_at?: string | null;
};

export type BusinessCostCenter = {
  cost_center_id: string;
  workspace_id: string;
  name: string;
  budget_limit: string | number | null;
  created_at?: string | null;
};

export type BusinessMember = {
  member_id: string;
  workspace_id: string;
  user_id: string;
  role: string;
  unit_id: string | null;
  created_at?: string | null;
};

export type BusinessMemberInvite = {
  invite_id: string;
  workspace_id: string;
  email: string;
  role: string;
  unit_id: string | null;
  accepted_at?: string | null;
  expires_at?: string | null;
  created_at?: string | null;
};

export type BusinessInvitePreview = {
  workspace_id: string;
  workspace_title: string;
  email: string;
  role: string;
  unit_id: string | null;
  unit_name?: string | null;
};

export type BusinessVendor = {
  vendor_id: string;
  workspace_id: string;
  name: string;
  vendor_type: string;
  contact_info: string | null;
  created_at?: string | null;
};

export type BusinessSpend = {
  spend_id: string;
  workspace_id: string;
  unit_id: string;
  title: string;
  amount: string | number;
  price_per_unit?: string | number | null;
  quantity?: string | number | null;
  measurement_unit?: string | null;
  spend_type: string;
  cost_center_id: string | null;
  vendor_id: string | null;
  status: "pending" | "approved" | "rejected" | string;
  submitted_by: string;
  approved_by: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  rejection_reason?: string | null;
  created_at?: string | null;
};

export type BusinessSignal = {
  signal_id: string;
  workspace_id: string;
  unit_id: string | null;
  signal_type: string;
  severity: "info" | "warning" | "critical" | string;
  title?: string | null;
  message: string;
  action_type?: string | null;
  action_target_type?: string | null;
  action_target_id?: string | null;
  resolved: boolean;
  created_at?: string | null;
};

export type BusinessActivity = {
  activity_id: string;
  workspace_id: string;
  unit_id: string | null;
  spend_id: string | null;
  actor_id: string | null;
  event_type: string;
  message: string;
  created_at?: string | null;
};

export type BusinessDashboard = {
  summary: {
    total_budget: string | number | null;
    total_spent: string | number;
    pending_amount: string | number;
    remaining: string | number | null;
  };
  pending_approvals: BusinessSpend[];
  approved_spends: BusinessSpend[];
  cost_center_breakdown: {
    key: string;
    label: string;
    amount: string | number;
    budget_limit: string | number | null;
    utilization_ratio: number | null;
  }[];
  unit_breakdown: {
    key: string;
    label: string;
    amount: string | number;
    budget_limit: string | number | null;
    utilization_ratio: number | null;
  }[];
  signals: BusinessSignal[];
  activity: BusinessActivity[];
};

export async function fetchBusinessWorkspaces(
  token: string,
  includeArchived = false,
): Promise<BusinessWorkspace[]> {
  const q = includeArchived ? "?include_archived=true" : "";
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces${q}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessWorkspace[]>(res);
}

export async function createBusinessWorkspace(
  token: string,
  body: {
    title: string;
    business_type: string;
    total_budget?: number | null;
    currency?: string;
  },
): Promise<BusinessWorkspace> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessWorkspace>(res);
}

export async function fetchBusinessWorkspace(
  token: string,
  workspaceId: string,
): Promise<BusinessWorkspace> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessWorkspace>(res);
}

export async function fetchBusinessUnits(token: string, workspaceId: string): Promise<BusinessUnit[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/units`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessUnit[]>(res);
}

export async function createBusinessUnit(
  token: string,
  workspaceId: string,
  body: {
    name: string;
    unit_type: string;
    location?: string | null;
    manager_user_id?: string | null;
    budget_limit?: number | null;
  },
): Promise<BusinessUnit> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/units`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessUnit>(res);
}

export async function fetchBusinessMembers(token: string, workspaceId: string): Promise<BusinessMember[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/members`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessMember[]>(res);
}

export async function createBusinessMember(
  token: string,
  workspaceId: string,
  body: {
    user_id: string;
    role: string;
    unit_id?: string | null;
  },
): Promise<BusinessMember> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/members`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessMember>(res);
}

export async function updateBusinessMember(
  token: string,
  workspaceId: string,
  memberId: string,
  body: {
    role?: string;
    unit_id?: string | null;
  },
): Promise<BusinessMember> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/members/${memberId}`, {
    method: "PATCH",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessMember>(res);
}

export async function inviteBusinessMemberByEmail(
  token: string,
  workspaceId: string,
  body: { email: string; role: string; unit_id?: string | null },
): Promise<{
  invite_id: string;
  workspace_id: string;
  email: string;
  role: string;
  unit_id: string | null;
  token?: string | null;
  join_url: string;
  sent: boolean;
  message?: string | null;
}> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/invites`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson(res);
}

export async function fetchBusinessMemberInvites(
  token: string,
  workspaceId: string,
): Promise<BusinessMemberInvite[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/invites`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessMemberInvite[]>(res);
}

export async function fetchBusinessInvitePreview(token: string): Promise<BusinessInvitePreview> {
  const res = await fetch(`${getApiBaseUrl()}/business/invites/preview?token=${encodeURIComponent(token)}`);
  return parseJson<BusinessInvitePreview>(res);
}

export async function acceptBusinessInvite(
  authToken: string,
  token: string,
): Promise<{ workspace_id: string; member_id: string; role: string; unit_id: string | null }> {
  const res = await fetch(`${getApiBaseUrl()}/business/invites/accept`, {
    method: "POST",
    headers: await headersJson(authToken),
    body: JSON.stringify({ token }),
  });
  return parseJson(res);
}

export async function fetchBusinessCostCenters(
  token: string,
  workspaceId: string,
): Promise<BusinessCostCenter[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/cost-centers`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessCostCenter[]>(res);
}

export async function createBusinessCostCenter(
  token: string,
  workspaceId: string,
  body: { name: string; budget_limit?: number | null },
): Promise<BusinessCostCenter> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/cost-centers`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessCostCenter>(res);
}

export async function fetchBusinessVendors(token: string, workspaceId: string): Promise<BusinessVendor[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/vendors`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessVendor[]>(res);
}

export async function createBusinessVendor(
  token: string,
  workspaceId: string,
  body: { name: string; vendor_type: string; contact_info?: string | null },
): Promise<BusinessVendor> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/vendors`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessVendor>(res);
}

export async function fetchBusinessSpends(token: string, workspaceId: string): Promise<BusinessSpend[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/spends`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessSpend[]>(res);
}

export async function submitBusinessSpend(
  token: string,
  workspaceId: string,
  body: {
    unit_id: string;
    title: string;
    amount?: number;
    price_per_unit?: number;
    quantity?: number;
    measurement_unit?: string | null;
    spend_type?: string;
    cost_center_id?: string | null;
    vendor_id?: string | null;
  },
): Promise<BusinessSpend> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/spends`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<BusinessSpend>(res);
}

export async function approveBusinessSpend(token: string, spendId: string): Promise<BusinessSpend> {
  const res = await fetch(`${getApiBaseUrl()}/business/spends/${spendId}/approve`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessSpend>(res);
}

export async function rejectBusinessSpend(
  token: string,
  spendId: string,
  reason: string,
): Promise<BusinessSpend> {
  const res = await fetch(`${getApiBaseUrl()}/business/spends/${spendId}/reject`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify({ reason }),
  });
  return parseJson<BusinessSpend>(res);
}

export async function fetchBusinessDashboard(
  token: string,
  workspaceId: string,
): Promise<BusinessDashboard> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/dashboard`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessDashboard>(res);
}

export async function fetchBusinessSignals(token: string, workspaceId: string): Promise<BusinessSignal[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/signals`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessSignal[]>(res);
}

/** GET /business/workspaces/{id}/today */
export type BusinessTodayCard = {
  signal_type: string;
  severity: string;
  title: string;
  message: string;
  action_type: string;
  action_target_type: string;
  action_target_id: string | null;
};

export type BusinessToday = {
  top_cards: BusinessTodayCard[];
  spend_today_amount: string | number;
  spends_submitted_today: number;
  approvals_handled_today: number;
  daily_status: string;
  daily_status_label: string;
};

export async function fetchBusinessToday(token: string, workspaceId: string): Promise<BusinessToday> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/today`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessToday>(res);
}

/** GET /business/workspaces/{id}/control-summary */
export type BusinessControlSummary = {
  total_budget: string | number | null;
  approved_spend: string | number;
  pending_spend: string | number;
  remaining_budget: string | number | null;
  approvals_count: number;
  control_score: number | null;
  control_label: string;
};

export async function fetchBusinessControlSummary(
  token: string,
  workspaceId: string,
): Promise<BusinessControlSummary> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/control-summary`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessControlSummary>(res);
}

/** GET /business/workspaces/{id}/recommendations */
export type BusinessRecommendation = {
  recommendation_id?: string | null;
  workspace_id: string;
  recommendation_type: string;
  priority: number;
  title: string;
  message: string;
  action_type: string;
  action_target_type: string;
  action_target_id: string | null;
  created_at?: string | null;
};

export async function fetchBusinessRecommendations(
  token: string,
  workspaceId: string,
): Promise<BusinessRecommendation[]> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/recommendations`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessRecommendation[]>(res);
}

/** GET /business/workspaces/{id}/insights */
export type BusinessInsights = {
  trend_weekly_pct: number | null;
  trend_weekly_label: string;
  spend_today_amount: string | number;
  top_unit_week_label: string | null;
  top_vendor_week_label: string | null;
  top_vendor_week_share_pct: number | null;
  monthly_budget_usage_pct: number | null;
  monthly_approvals_count: number;
  chips: string[];
};

export async function fetchBusinessInsights(token: string, workspaceId: string): Promise<BusinessInsights> {
  const res = await fetch(`${getApiBaseUrl()}/business/workspaces/${workspaceId}/insights`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<BusinessInsights>(res);
}

File: backend/app/schemas/business.py
from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class BusinessWorkspaceCreate(BaseModel):
    title: str = Field(max_length=255)
    business_type: str = Field(max_length=64)
    total_budget: Decimal | None = None
    currency: str = Field(default="INR", max_length=8)
    status: str = Field(default="active", max_length=16)


class BusinessWorkspaceUpdate(BaseModel):
    title: str | None = Field(default=None, max_length=255)
    business_type: str | None = Field(default=None, max_length=64)
    total_budget: Decimal | None = None
    currency: str | None = Field(default=None, max_length=8)
    status: str | None = Field(default=None, max_length=16)


class BusinessWorkspaceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    workspace_id: UUID
    title: str
    business_type: str
    total_budget: Decimal | None
    currency: str
    created_by: str
    status: str
    created_at: str | None = None
    updated_at: str | None = None


class BusinessUnitCreate(BaseModel):
    name: str = Field(max_length=160)
    unit_type: str = Field(max_length=32)
    location: str | None = Field(default=None, max_length=255)
    manager_user_id: str | None = Field(default=None, max_length=128)
    budget_limit: Decimal | None = None
    status: str = Field(default="active", max_length=16)


class BusinessUnitOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    unit_id: UUID
    workspace_id: UUID
    name: str
    unit_type: str
    location: str | None
    manager_user_id: str | None
    budget_limit: Decimal | None
    status: str
    created_at: str | None = None


class BusinessMemberCreate(BaseModel):
    user_id: str = Field(max_length=128)
    role: str = Field(max_length=24)
    unit_id: UUID | None = None


class BusinessMemberUpdate(BaseModel):
    role: str | None = Field(default=None, max_length=24)
    unit_id: UUID | None = None


class BusinessMemberOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    member_id: UUID
    workspace_id: UUID
    user_id: str
    role: str
    unit_id: UUID | None
    created_at: str | None = None


class BusinessInviteSendBody(BaseModel):
    email: str = Field(min_length=5, max_length=320)
    role: str = Field(max_length=24)
    unit_id: UUID | None = None


class BusinessInviteSendOut(BaseModel):
    invite_id: UUID
    workspace_id: UUID
    email: str
    role: str
    unit_id: UUID | None = None
    token: str | None = None
    join_url: str
    sent: bool
    message: str | None = None
    created_at: str | None = None


class BusinessInvitePreviewOut(BaseModel):
    workspace_id: UUID
    workspace_title: str
    email: str
    role: str
    unit_id: UUID | None = None
    unit_name: str | None = None


class BusinessInviteAcceptBody(BaseModel):
    token: str = Field(min_length=8, max_length=255)


class BusinessInviteAcceptOut(BaseModel):
    workspace_id: UUID
    member_id: UUID
    role: str
    unit_id: UUID | None = None


class BusinessInviteOut(BaseModel):
    invite_id: UUID
    workspace_id: UUID
    email: str
    role: str
    unit_id: UUID | None = None
    accepted_at: str | None = None
    expires_at: str | None = None
    created_at: str | None = None


class BusinessCostCenterCreate(BaseModel):
    name: str = Field(max_length=160)
    budget_limit: Decimal | None = None


class BusinessCostCenterOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    cost_center_id: UUID
    workspace_id: UUID
    name: str
    budget_limit: Decimal | None
    created_at: str | None = None


class BusinessVendorCreate(BaseModel):
    name: str = Field(max_length=200)
    vendor_type: str = Field(max_length=64)
    contact_info: str | None = None


class BusinessVendorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    vendor_id: UUID
    workspace_id: UUID
    name: str
    vendor_type: str
    contact_info: str | None
    created_at: str | None = None


class BusinessSpendCreate(BaseModel):
    unit_id: UUID
    title: str = Field(max_length=255)
    amount: Decimal | None = None
    price_per_unit: Decimal | None = None
    quantity: Decimal | None = None
    measurement_unit: str | None = Field(default=None, max_length=32)
    spend_type: str = Field(default="operational", max_length=64)
    cost_center_id: UUID | None = None
    vendor_id: UUID | None = None


class BusinessSpendOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    spend_id: UUID
    workspace_id: UUID
    unit_id: UUID
    title: str
    amount: Decimal
    price_per_unit: Decimal | None = None
    quantity: Decimal | None = None
    measurement_unit: str | None = None
    spend_type: str
    cost_center_id: UUID | None
    vendor_id: UUID | None
    status: str
    submitted_by: str
    approved_by: str | None
    submitted_at: str | None = None
    approved_at: str | None = None
    rejection_reason: str | None = None
    created_at: str | None = None


class BusinessRejectBody(BaseModel):
    reason: str = Field(min_length=1, max_length=2000)


class BusinessSignalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    signal_id: UUID
    workspace_id: UUID
    unit_id: UUID | None
    signal_type: str
    severity: str
    title: str | None = None
    message: str
    action_type: str | None = None
    action_target_type: str | None = None
    action_target_id: str | None = None
    resolved: bool
    created_at: str | None = None


class BusinessRecommendationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    recommendation_id: UUID | None = None
    workspace_id: UUID
    recommendation_type: str
    priority: int
    title: str
    message: str
    action_type: str
    action_target_type: str
    action_target_id: str | None = None
    created_at: str | None = None


class BusinessControlSummaryOut(BaseModel):
    total_budget: Decimal | None
    approved_spend: Decimal
    pending_spend: Decimal
    remaining_budget: Decimal | None
    approvals_count: int
    control_score: int | None = None
    control_label: str


class BusinessTodayCardOut(BaseModel):
    signal_type: str
    severity: str
    title: str
    message: str
    action_type: str
    action_target_type: str
    action_target_id: str | None = None


class BusinessTodayOut(BaseModel):
    top_cards: list[BusinessTodayCardOut] = Field(default_factory=list)
    spend_today_amount: Decimal
    spends_submitted_today: int
    approvals_handled_today: int
    daily_status: str
    daily_status_label: str


class BusinessUnitPerformanceOut(BaseModel):
    key: str
    label: str
    unit_type: str | None = None
    amount: Decimal
    pending_amount: Decimal | None = None
    budget_limit: Decimal | None = None
    utilization_ratio: float | None = None
    utilization_pct: float | None = None
    performance_state: str


class BusinessUnitPerformanceResponseOut(BaseModel):
    rows: list[BusinessUnitPerformanceOut] = Field(default_factory=list)
    top_unit_id: str | None = None
    top_unit_label: str | None = None
    at_risk_unit_id: str | None = None
    at_risk_unit_label: str | None = None


class BusinessCostCenterBreakdownOut(BaseModel):
    key: str
    label: str
    amount: Decimal
    pending_amount: Decimal | None = None
    budget_limit: Decimal | None = None
    utilization_ratio: float | None = None
    utilization_pct: float | None = None
    state: str


class BusinessInsightsOut(BaseModel):
    trend_weekly_pct: float | None = None
    trend_weekly_label: str
    spend_today_amount: Decimal
    top_unit_week_label: str | None = None
    top_vendor_week_label: str | None = None
    top_vendor_week_share_pct: float | None = None
    monthly_budget_usage_pct: float | None = None
    monthly_approvals_count: int
    chips: list[str] = Field(default_factory=list)


class BusinessActivityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    activity_id: UUID
    workspace_id: UUID
    unit_id: UUID | None
    spend_id: UUID | None
    actor_id: str | None
    event_type: str
    message: str
    created_at: str | None = None


class BusinessSummaryOut(BaseModel):
    total_budget: Decimal | None
    total_spent: Decimal
    pending_amount: Decimal
    remaining: Decimal | None


class BusinessBreakdownRowOut(BaseModel):
    key: str
    label: str
    amount: Decimal
    budget_limit: Decimal | None = None
    utilization_ratio: float | None = None


class BusinessDashboardOut(BaseModel):
    summary: BusinessSummaryOut
    pending_approvals: list[BusinessSpendOut] = Field(default_factory=list)
    approved_spends: list[BusinessSpendOut] = Field(default_factory=list)
    cost_center_breakdown: list[BusinessBreakdownRowOut] = Field(default_factory=list)
    unit_breakdown: list[BusinessBreakdownRowOut] = Field(default_factory=list)
    signals: list[BusinessSignalOut] = Field(default_factory=list)
    activity: list[BusinessActivityOut] = Field(default_factory=list)

